tasty-query/shared/src/main/scala/tastyquery/Symbols.scala

```diff
@@ -417,6 +417,78 @@ object Symbols {
       */
     final def isStableMember(using Context): Boolean =
       !isAnyOf(Method | Mutable) && !declaredType.isInstanceOf[ExprType]
+
+    /** The symbol actually implementing this symbol for an object of class `runtimeClass`.
+      *
+      * If this symbol is private or if it is a constructor,
+      * `runtimeImplementingSymbol` returns this symbol.
+      *
+      * Otherwise, it returns the first non-abstract symbol matching this
+      * symbol in the linearization of `runtimeClass`.
+      *
+      * @throws java.lang.IllegalArgumentException
+      *   if `this.owner.isClass` is false or if
+      *   `runtimeClass.isSubclass(this.owner.asClass)` is false
+      * @throws tastyquery.Exceptions.MemberNotFoundException
+      *   if no concrete matching symbol could be found
+      */
+    final def runtimeImplementingSymbol(runtimeClass: ClassSymbol)(using Context): TermSymbol =
+      if canMatchInheritedSymbols then
+        val linearization = runtimeClass.linearization
+        if !linearization.contains(this.owner.asClass) then
+          throw IllegalArgumentException(s"Invalid run-time class $runtimeClass for dynamic super call to $this")
+        findRuntimeSymbolCommon(runtimeClass, linearization)
+      else
+        if !this.owner.isClass then
+          throw IllegalArgumentException(s"Nonsensical run-time implementation lookup for local symbol $this")
+        this
+    end runtimeImplementingSymbol
+
+    /** Resolves a dynamically bound `super` call for an object of class `runtimeClass`.
+      *
+      * This method returns the next symbol for a `super` call of the form
+      * `enclosingClass.super.foo` for an object of class `runtimeClass`.
```

Comment:
`enclosingClass.super.foo` is confusing because in Scala, inside a class `A`, you cannot refer to a _parent_ class `B` as `B.super.foo`; you can only use that to refer to an _enclosing_ class in the sense that `A` is an inner class in `B`. But in tasty-query both of those things are represented the same way, as the type of the `This` inside the `Super`. And for this method in particular, we do require `enclosingClass` to be a parent of `runtimeClass` and definitely not an outer class.

Reply:
I notice that the same `B.super.foo` notation is used in the docstring of `Super` too, perhaps it should be clarified there too. And maybe we should add a test for `nextSuperSymbol` that actually uses an inner class and demonstrates that the linearization of the inner class does not matter when resolving `Outer.super.foo` and vice versa.

Reply:
All you say is true, but I don't see why it contradicts the use of `enclosingClass` here. Precisely, it must be enclosing the `super` call for it to be valid. Now tasty-query has no notion of "where" we "are", so we can give it an enclosing class and tasty-query has to trust that it is meaningful.

For the same reason, a test with an inner class would not change anything. We would pass exactly the same arguments to this method whether we "are" in the inner class or not. As far as a test goes, we would be copying the same thing twice without any difference.

Reply:
I entirely agree, I’m not saying it contradicts the use of `enclosingClass`. I’m saying it contradicts the use of `enclosingClass.super.foo`. Because if you have a class `B` with an inner class `A`, then `B` is a valid "enclosing class" of a `super` in the body of `A` which we could write as `B.super.foo`. But to evaluate `B.super.foo`, the runtime class must inherit from `B`, not from `A`, right? I’m just saying that writing it as `enclosingClass.super.foo` is confusing because it is not sufficient for `enclosingClass` to be an "enclosing class" of the `super`, but in addition it must be a superclass of the `runtimeClass`, which may not necessarily inherit from the class directly enclosing the `super` (`A` in our example).